url147ce07ebe_JoshuaKissoon_Kademlia_tgz-pJ8-kademlia_simulations_RoutingTableStateTestingJ8/src/kademlia/JKademliaNode.java

```diff
@@ -0,0 +1,429 @@
+        return this.dht;
+    }
+
+    @Override
+    public KadConfiguration getCurrentConfiguration()
+    {
+        return this.config;
+    }
+
+    @Override
+    public synchronized final void bootstrap(Node n) throws IOException, RoutingException
+    {
+        long startTime = System.nanoTime();
+        Operation op = new ConnectOperation(this.server, this, n, this.config);
+        op.execute();
+        long endTime = System.nanoTime();
+        this.statistician.setBootstrapTime(endTime - startTime);
+    }
+
+    @Override
+    public int put(KadContent content) throws IOException
+    {
+        return this.put(new JKademliaStorageEntry(content));
+    }
+
+    @Override
+    public int put(JKademliaStorageEntry entry) throws IOException
+    {
+        StoreOperation sop = new StoreOperation(this.server, this, entry, this.dht, this.config);
+        sop.execute();
+
+        /* Return how many nodes the content was stored on */
+        return sop.numNodesStoredAt();
+    }
+
+    @Override
+    public void putLocally(KadContent content) throws IOException
+    {
+        this.dht.store(new JKademliaStorageEntry(content));
+    }
+
+    @Override
+    public JKademliaStorageEntry get(GetParameter param) throws NoSuchElementException, IOException, ContentNotFoundException
+    {
+        if (this.dht.contains(param))
+        {
+            /* If the content exist in our own DHT, then return it. */
+            return this.dht.get(param);
+        }
+
+        /* Seems like it doesn't exist in our DHT, get it from other Nodes */
+        long startTime = System.nanoTime();
+        ContentLookupOperation clo = new ContentLookupOperation(server, this, param, this.config);
+        clo.execute();
+        long endTime = System.nanoTime();
+        this.statistician.addContentLookup(endTime - startTime, clo.routeLength(), clo.isContentFound());
+        return clo.getContentFound();
+    }
+
+    @Override
+    public void refresh() throws IOException
+    {
+        new KadRefreshOperation(this.server, this, this.dht, this.config).execute();
+    }
+
+    @Override
+    public String getOwnerId()
+    {
+        return this.ownerId;
+    }
+
+    @Override
+    public int getPort()
+    {
+        return this.udpPort;
+    }
+
+    @Override
+    public void shutdown(final boolean saveState) throws IOException
+    {
+        /* Shut down the server */
+        this.server.shutdown();
+
+        this.stopRefreshOperation();
+
+        /* Save this Kademlia instance's state if required */
+        if (saveState)
+        {
+            /* Save the system state */
+            this.saveKadState();
+        }
+    }
+
+    @Override
+    public void saveKadState() throws IOException
+    {
+        DataOutputStream dout;
+
+        /**
+         * @section Store Basic Kad data
+         */
+        dout = new DataOutputStream(new FileOutputStream(getStateStorageFolderName(this.ownerId, this.config) + File.separator + "kad.kns"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *dout*. Currently, there are execution paths that do not contain closure statements, for example, when JsonSerializer.write throws an exception. Ensure *dout* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).